Filesystem Hierarchy and Packaging Proposed Modifications

rossmcdonald opened the pull request:
## Summary

This PR is meant to outline some proposed changes to the InfluxDB file system hierarchy and packaging.

### Packaging

Changes to packaging:

* Updated logrotate configuration has been updated from `/etc/logrotate.d/influxd` to `/etc/logrotate.d/influxdb` (the legacy file name will be removed on install). The file permissions have also been modified to be `644`. #4446 #4447 
* The `influxdb` user's shell is now set to `/bin/false`. #3753 
* The InfluxDB init and service scripts are copied (not symlinked) to their appropriate locations from `/usr/lib/influxdb/scripts`. #3753
* The `/etc/default/influxdb` file is now created on install. #3753

### Filesystem Hierarchy

Changes to the filesystem hierarchy:

* InfluxDB binaries are now installed to `/usr/bin`. #3753
* InfluxDB configuration now lives in `/etc/influxdb`. #3753 #3123 
* InfluxDB data now lives in `/var/lib/influxdb`. #3123 
* InfluxDB scripts now live in `/usr/lib/influxdb/scripts`

Please let me know if there are any objections to the above changes.

jwilder commented:
Location changes look good.

Needs a changelog and release notes entry as this will likely affect people using config management tools.

otoolep commented:
What is an example of something that goes into `/usr/lib/influxdb/scripts`?

otoolep commented:
Oh, the init script. I thought this would just be dropped into `/etc` directly, no?

otoolep commented on package.sh:
```diff
@@ -387,13 +385,13 @@ do
-            echo
-            usage 1
-        fi
+        # echo $VERSION | grep -i '[r|c]' 2>&1 >/dev/null
```
Do you mean to comment this out?

rossmcdonald replied:
Fixed.

otoolep commented:
Can you rebase to remove merge commits?

rossmcdonald commented:
> Oh, the init script. I thought this would just be dropped into /etc directly, no?

The post-install script has extra logic to determine whether the system is using systemd or sysvinit, so I figured it would be easier to just copy them over. Otherwise we will need unique builds for systemd distributions vs sysvinit distributions.

otoolep commented on package.sh:
```diff
@@ -213,9 +214,11 @@ check_tag_exists () {
     version=$2
-    mkdir -p $work_dir/$INSTALL_ROOT_DIR/versions/$version/scripts
+
+    mkdir -p $work_dir/$INSTALL_ROOT_DIR
```
What about checking an error here? Just to be super careful.

otoolep replied:
We do plenty of `mkdir` later on without checking codes. Perhaps don't bother checking any? Not sure.

rossmcdonald replied:
I agree we're not consistent in our checks. I'm going to be doing a larger revision to the package script to allow for multi-arch builds, so I'll work on improving that with the next batch of changes.

otoolep replied:
Works for me.

otoolep commented:
Looks great to me. Once this merged I'd like to invite everyone to a ceremonial burning of the FHS, since no-one ever seems to follow it -- except yours truly.

Managing this change will be important. A clear section titled "Release Notes" is required in the CHANGELOG.

otoolep commented:
Can you outline the upgrade testing you performed? Did you upgrade from 0.9.4, and then move existing data and config files? Or something else?

rossmcdonald commented:
Right now the upgrade path will be (for a < 0.9.5 instance):
* Install the v0.9.5 package, stopping the service is not required
* Manually move `/var/opt/influxdb/*` to `/var/lib/influxdb/*` (the directory will already exist)
* Manually copy the `/etc/opt/influxdb/influxdb.conf` file to `/etc/influxdb/influxdb.conf` (or just modify the existing configuration)

To test I:
* Installed InfluxDB 0.9.4.2, loaded test data
* Installed InfluxDB v0.9.5 without stopping the service (verifying that a `service influxdb status` still shows the running process)
* Stop the InfluxDB process with a `service influxdb stop`
* Move everything in `/var/opt/influxdb` to `/var/lib/influxdb`
* Modify any relevant parts of `/etc/influxdb/influxdb.conf` as needed (really just the `engine`)
* Start the InfluxDB service with a `service influxdb start`
* Verify that the previous set of data is there, and can be queried successfully
* Restart service to ensure no other issues

The problems I think most people will hit:
* if they auto-upgrade their instance, then when the service is restarted it will automatically try to use the new directories (making it look like there's no data in the database). 
* copy the configuration from `/etc/opt/influxdb/influxdb.conf` to `/etc/influxdb`, overwriting the new configuration (resetting the directories and storage engine to the legacy ones)

I can also create an upgrade script (in `/usr/lib/influxdb`) that could perform the migration automatically if you think there may be issues with forcing users to do it manually.

rossmcdonald commented:
@gunnaraasen @beckettsean What do you guys think of the potential upgrade path above? Is manually moving the data and configuration directories too much to ask?

otoolep commented:
OK, this testing seems fine to me.

I think the release notes should include a link to the process outlined above, so it's clear what people need to do. I would rather take this hit as soon as possible.

beckettsean commented:
I think the best way to handle it would be as part of the blog post about the new packaging structure. This would be a section at the end of the post, describing how to do the upgrade.

beckettsean commented:
To clarify, release notes are an excellent place for it, too, but I think the blog post is more likely to jump out from Google searches than the CHANGELOG.

otoolep commented:
Blog post works for me -- brief release notes, and reference the post.

gunnaraasen commented:
Not sure if this was already discussed, but is there a reason we can't or don't want to automatically move `/var/opt/influxdb/*` to `/var/lib/influxdb/*`?

rossmcdonald commented:
I originally considered adding that to the post-installation script, but the possibility of something going terribly wrong was just too high. Mainly:
* What if `/var/opt` and `/var/lib` are on two different drives? How do we know we have enough space on `/var/lib`? And if they have a ton of data, the move might be a pretty time-consuming operation.
* What if the service isn't stopped correctly because of someone's weird setup (or they have another daemon that attempts restarts)? Attempting to move the directories automatically could corrupt data and break the server.
* If they are using a highly customized configuration, the chance of accidentally overwriting/modifying something is pretty high.

All of these could be handled in the post-inst script, but the permutations of different configurations is just so high. In the end I figured it would just be better to make this a manual step. The worst-case scenario with making it manual is that the admin forgot to move the directories over, and it appears as if the database is empty (since the configuration will automatically point to the new directories, but the data lives in the old ones).

otoolep commented:
Agreed. In principle it's a good idea @gunnaraasen, but all the edge cases outlined by @rossmcdonald argue strongly against it.

Post-install scripts for tweaking a config file is OK, but moving data around is a much bigger issue.

gunnaraasen commented:
Makes sense. 

How about adding a post-install script to check `/var/opt/influxdb/*` and print a big warning message if it's not empty?

We can also make a note about upgrading on the downloads page and point to the blog post.

rossmcdonald commented:
> How about adding a post-install script to check /var/opt/influxdb/* and print a big warning message if it's not empty?

That's a good idea. I can add in some scary messages to the post-inst script letting the user know they need to move everything over and update their configuration.

otoolep commented:
I am OK with adding this warning, but eventually (like, say around 1.0) I'd nuke it. IMHO.

nathanielc commented:
Adding in my two cents:

When I upgrade a package I expect:

* The service still works on restart
* The upgrade doesn't change my config, but rather drops the equivalent of config.rpmnew in the correct location.
* Does not try to migrate any data etc.

We could just look for the old config at `/etc/opt/influxdb/influxdb.conf` and if we find it copy it to `/etc/influxdb/influxdb.conf` as a preinstall step, then when the new package is installed it will see a differing config file and take the appropriate action (in the case of an rpm, it will leave the config alone and create a new file ending in .rpmnew, not sure how deb behaves). By coping the old config into the new location we can ensure that the service restart will work since it will still look for the data in the old locations. At that point its up to the sysadmin to check for new configuration and migrate the data according to his or her needs.

nathanielc commented on scripts/influxdb.service:
```diff
@@ -9,9 +9,11 @@ After=network.target
 EnvironmentFile=-/etc/default/influxdb
-ExecStart=/opt/influxdb/influxd -config /etc/opt/influxdb/influxdb.conf $INFLUXD_OPTS
-KillMode=process
+ExecStart=/bin/sh -c "/usr/bin/influxd -config /etc/influxdb/influxdb.conf ${INFLUXD_OPTS} > ${STDOUT} 2> ${STDERR}"
```
We should append the output so logs aren't truncated on a restart.
`ExecStart=/bin/sh -c "/usr/bin/influxd -config /etc/influxdb/influxdb.conf ${INFLUXD_OPTS} > > ${STDOUT} 2>> ${STDERR}"`

nathanielc replied:
Should we even do this right now? Seems like if the binary can't log to a file specifically then we should just use the journal. Not a really big deal but doing something weird like this can often cause confusion instead of being helpful.

otoolep replied:
Good catch -- append the logs.

rossmcdonald replied:
Updated.

rossmcdonald commented:
Updated to include the most recent changes:
* Per @nathanielc, added a pre-install script to check for existence of `/etc/opt/influxdb/influxdb.conf`. If found, it will copy the directory to `/etc/influxdb`. The default behavior of the package manager is then to drop the new config in the same directory with a different name, and leave the existing one alone. This ensures the directories and settings are all set to the same paths as they were before. They will still be using the old directories, but at least nothing will break when restarted.
* For systemd installs, we are still using the hacky method of redirecting the logs by wrapping the process in `sh`. I'd really prefer to have something built-in to InfluxDB that can be used to specify the log files through configuration, but I think for the time being this should work well enough. My main concern is around supportability, but let me know if anyone thinks otherwise.
* I've also added a new build script (`build.py`) which can be used to generate builds and packages for any/all support platforms and architectures at once, eventually replacing `package.sh`. Python is just so much more flexible, and it has no external dependencies apart from `fpm` and `boto` for packaging and S3 uploads (which are not required for builds). I've also created two other similar scripts for Telegraf and *friend*, so that all Influx projects can be built in a near-identical manner.

Let me know if anyone has any other thoughts or feedback!

otoolep commented:
All onboard for `build.py`. bash can only get you so far.

otoolep commented:
+1, great work.

nathanielc commented:
+1 

@otoolep bash **should** only get you so far ;)

otoolep commented:
Whoop-de-do-da.
